packages/convex-helpers/server/retries.ts

```diff
@@ -0,0 +1,171 @@
+/**
+ * This file defines helper functions that can be used to retry a
+ * Convex action until it succeeds. An action should only be retried if it is
+ * safe to do so, i.e., if it's idempotent or doesn't have any unsafe side effects.
+ *
+ * Used like:
+ * ```ts
+ * // in convex/utils.ts
+ * import { makeActionRetrier } from "convex-helpers/server/retries";
+ * import { internalMutation } from "./convex/_generated/server";
+ *
+ * export const { runWithRetries, retry } = makeActionRetrier(internalMutation, internal.utils.retry);
+ *
+ * // in a mutation or action
+ * await runWithRetries(ctx, internal.myModule.myAction, { arg1: 123 });
+ * ```
+ */
+import {
+  FunctionReference,
+  FunctionVisibility,
+  Scheduler,
+  FunctionArgs,
+  OptionalRestArgs,
+  getFunctionName,
+  makeFunctionReference,
+  DefaultFunctionArgs,
+  internalMutationGeneric,
+} from "convex/server";
+import { v, ObjectType } from "convex/values";
+
+const DEFAULT_WAIT_BACKOFF = 10;
+const DEFAULT_RETRY_BACKOFF = 10;
+const DEFAULT_BASE = 2;
+const DEFAULT_MAX_FAILURES = 16;
+const retryArguments = {
+  job: v.id("_scheduled_functions"),
+  action: v.string(),
+  actionArgs: v.any(),
+  waitBackoff: v.number(),
+  retryBackoff: v.number(),
+  base: v.number(),
+  maxFailures: v.number(),
+};
+
+/**
+ * Create a function that retries an action with exponential backoff.
+ * e.g.
+ * ```ts
+ * // in convex/utils.ts
+ * import { makeActionRetrier } from "convex-helpers/server/retries";
+ * import { internalMutation } from "./convex/_generated/server";
+ *
+ * export const { runWithRetries, retry } = makeActionRetrier(internalMutation, internal.utils.retry);
+ *
+ * // in a mutation or action
+ * await runWithRetries(ctx, internal.myModule.myAction, { arg1: 123 });
+ * ```
+ *
+ * @param internalMutation From "./convex/_generated/server" or customMutation.
+ * @param retryRef The function reference to the retryRef function exported.
+ * e.g. internal.mymodule.retry
+ * @returns An object with runWithRetries and retry functions.
+ */
+export function makeActionRetrier(retryFnName: string) {
```

Comment:
It would be great to have a stable reference to a function. 
Components would work well here - a well-known function to call.
This needs a string instead of the type-safe FunctionReference b/c it'd be circular types.